Pin connector refactor

messmerd opened the pull request:
Major refactor of audio plugin implementations to add support for the pin connector to MIDI-based instrument plugins and all effect plugins. Also refactors parts of the pin connector itself.

The idea was to add a new class template called `AudioPluginInterface` which audio plugins can inherit from rather than inheriting directly from `Instrument` or `Effect`. This class template uses template parameters to allow compile-time customization for plugins (i.e. how many inputs, how many outputs, whether in-place processing is used, etc.) and is responsible for owning and using the pin connector and its routing capabilities. `AudioPluginInterface` is entirely optional - a plugin implementation could choose to continue inheriting from `Instrument` or `Effect` directly and all it would mean is no support for the pin connector.

One of the challenges along the way was how to deal with different working buffers used by plugins. The size, data type, and data layout differ depending on plugin characteristics such as how many in/out channels there are, whether in-place processing is used, etc. But also, VSTs and ZynAddSubFX cannot just use any buffer - they use a shared memory buffer.

The solution I came up with is to introduce `AudioPluginBufferInterface` which uses `AudioPluginInterface`'s compile-time customization options to customize its itself based on the needs of the plugin implementation. `AudioPluginBufferDefaultImpl` is provided which implements the interface in a way that works for most plugins. For VST and ZynAddSubFX, they specifically request (via the `customBuffer` option) to implement the buffer themselves and provide the audio plugin interface access to it through `AudioPluginBufferInterfaceProvider`. LV2, CLAP, Carla, and LADSPA could also take advantage of the custom buffers feature to reuse the same buffers they provide to their plugin APIs which would avoid extra memory and extra copying, though this work has not been done yet.

While this system is as flexible, simple, and performant as I could make it, it is not thread-safe which leads to a consequence I didn't foresee - `NotePlayHandle`-based instruments (which are all instruments except for Carla, CLAP, LV2, OpulenZ, Vestige, and ZynAddSubFX) are unable to use `AudioPluginInterface` because all the note play handles will use the same working buffer concurrently which messes up the audio. However, this isn't much of an issue since all `NotePlayHandle`-based instruments have exactly two fixed audio output channels, and the main pin connector use case for instruments is to provide support for a variable number of output channels or sidechain inputs. Pin connector support for `NotePlayHandle`-based instruments could potentially be added in the future, though it probably wouldn't be worth it.

This PR also adds unit tests for the pin connector.

messmerd committed:
Add AudioProcessor (WIP)

messmerd committed:
Fix build

messmerd committed:
Rename some audio data types

messmerd committed:
Introduce type for bus data

messmerd committed:
Add always_false_v

messmerd committed:
More work (WIP)

messmerd committed:
Update Span

Will want to replace with C++20's std::span as soon as possible

messmerd committed:
Templatize pin connector routing methods

messmerd committed:
Fix always_false_v

messmerd committed:
Add setFlag method to Flags

messmerd committed:
More progress (WIP)

messmerd committed:
Allow plugin implementations to provide their own working buffer

messmerd committed:
Span's non-const to const construction doesn't work

messmerd committed:
Implement pin connector routing for SampleFrame buffers

messmerd committed:
More progress (WIP)

messmerd committed:
Fix macOS builds

messmerd committed:
Fix narrowing conversion

messmerd committed:
Split the routing methods since it is actually possible to do

messmerd committed:
Rewrite routeFromPlugin implementation for non-SampleFrame

The new implementation does not require a working buffer and requires
much less copying, so it should be more efficient in terms of both space
and time

messmerd committed:
Fix some issues

messmerd committed:
Fix more issues

messmerd committed:
Progress (WIP)

messmerd committed:
Simplify Span

messmerd committed:
Major refactoring (WIP)

messmerd committed:
Fix build

messmerd committed:
Fix post-rebase mistake

messmerd committed:
Improve ergonomics using class type NTTP and designated initializers

messmerd committed:
Try to fix build

messmerd committed:
Fix some warnings on macOS

messmerd committed:
Try to fix Linux and MinGW builds

messmerd committed:
Add unit tests

messmerd committed:
Fix routeFromPlugin() for split audio data

messmerd committed:
Add more unit tests

messmerd committed:
Fix a few bugs

messmerd committed:
Remove NotePlayHandle-based instrument support from AudioPluginInterface

messmerd committed:
Add wet/dry processing (WIP)

messmerd committed:
Remove wet/dry mixing from plugin implementations (WIP)

messmerd committed:
Remove wet/dry mixing from Lv2Effect

messmerd committed:
Fix build

messmerd committed:
Revert "Fix build"

This reverts commit a013c43b2b9301ffaa2781ac50010b891de7e426.

messmerd committed:
Revert "Remove wet/dry mixing from Lv2Effect"

This reverts commit f40dd73fcefd8286f4cbfaaa2969abc528e2dd2d.

messmerd committed:
Revert "Remove wet/dry mixing from plugin implementations (WIP)"

This reverts commit 8a02c62101ef3ec628f60463236b96cb49f7f04e.

messmerd committed:
Revert "Add wet/dry processing (WIP)"

This reverts commit 4876d910ea43de97e80a4e58a7cf9e5a1160ace1.

messmerd committed:
Minor cleanup

JohannesLorenz commented:
Some developers had already proposed that using potentially different threads for different note play handles is overkill. This might even be changed when we make LMMS realtime safe. I assume in that case, thee instruments would be able to use `AudioPluginInterface`?

messmerd committed:
Remove pin connector from Carla, Lv2Instrument, OpulenZ

Pin connector support can be added in later. I think it's best to hold
off for now since Carla and Lv2 do not implement support for multiple
plugin channels yet, which could complicate upgrading project files
created between now and when that support is added.

Also cleaned up the plugin headers, removing some unecessary changes

messmerd commented:
@JohannesLorenz Yes, I'm not aware of anything that would prevent `AudioPluginInterface` from working with `NotePlayHandle`-based instruments if all the note play handles were to use the same thread.

JohannesLorenz commented on include/Instrument.h:
```diff
@@ -121,21 +131,21 @@ class LMMS_EXPORT Instrument : public Plugin
 	bool isMidiBased() const
 	{
-		return m_flags.testFlag(Instrument::Flag::IsMidiBased);
+		return !m_flags.testFlag(Instrument::Flag::IsMidiBased);
```
Either this change is wrong, or `Instrument::Flag::IsMidiBased` needs a description comment that it means the opposite of what its name suggests (or even better, the flag should be renamed).

messmerd replied:
Good catch - I'm not sure how this happened

messmerd replied:
Fixed in https://github.com/LMMS/lmms/pull/7459/commits/b81546ab2f43ce36fcee51cfb44904c5b01e4fa6

JohannesLorenz commented on plugins/Amplifier/Amplifier.h:
```diff
@@ -26,19 +26,19 @@
 {
 
-class AmplifierEffect : public Effect
+class AmplifierEffect : public DefaultEffectPluginInterface
```
The name `DefaultEffectPluginInterface` seems unnaturally long. "Effect" implicits "Plugin", and we usually don't differ between Interfaces and Non-Interfaces. Suggesting `DefaultEffect`.

messmerd replied:
I agree. I'll change it

messmerd replied:
Fixed in https://github.com/LMMS/lmms/pull/7459/commits/36c99d9263a2a9ae43dbc85c09d24c88803eaa4d

JohannesLorenz commented on plugins/Amplifier/Amplifier.h:
```diff
@@ -26,19 +26,19 @@
 #define LMMS_AMPLIFIER_H
 
-#include "Effect.h"
+#include "AudioPluginInterface.h"
```
Like in the previous comment: `AudioPlugin` would probably much more user friendly.

messmerd replied:
Fixed in https://github.com/LMMS/lmms/pull/7459/commits/36c99d9263a2a9ae43dbc85c09d24c88803eaa4d

JohannesLorenz commented on plugins/BassBooster/BassBooster.cpp:
```diff
@@ -86,9 +86,9 @@ Effect::ProcessStatus BassBoosterEffect::processImpl(SampleFrame* buf, const fpp
 	const float w = wetLevel();
 
-	for (fpp_t f = 0; f < frames; ++f)
+	for (fpp_t f = 0; f < inOut.size(); ++f)
```
Could we implement range-based `for` for `CorAudioDataMut` and friends? Like `for (fpp_t f : inOut.indices())`? (I am aware of the range-based `for` which does not carry the index, but if it is needed, like here, an index-variant might be practical and help avoiding dumb mistakes like `f <= inOut.size()`)

messmerd replied:
I'd rather not add any methods to `lmms::Span` that `std::span` does not have.

In C++23 we'll have `std::views::enumerate` which will provide a good solution to this problem:
```c++
for (auto&& [index, currentFrame] : std::views::enumerate(inOut))
{
	// ...
}
```

JohannesLorenz commented on plugins/CarlaBase/Carla.h:
```diff
@@ -195,7 +189,7 @@ class CARLABASE_EXPORT CarlaInstrument : public Instrument
     void saveSettings(QDomDocument& doc, QDomElement& parent) override;
     void loadSettings(const QDomElement& elem) override;
-    void play(SampleFrame* workingBuffer) override;
+    void playImpl(CoreAudioDataMut out) override;
```
Is there even a `play` function that would cause conflicts? If not, `play` would be shorter and possibly easier to understand for many (The name change also requires changing comments that referenced `play()`, e.g. the one above the `fMutex` variable.).

Even if this does conflict with `play` of a parent class, it might make more sense to rename that upper function and keep `play` in the final Plugin implementations consistent to the past.

It also looks strange because of asymmetry: saveSettings, loadSettings etc are also implementations, but not named `...Impl`.

JohannesLorenz commented on plugins/FreeBoy/FreeBoy.cpp:
```diff
@@ -228,7 +228,7 @@ float FreeBoyInstrument::desiredReleaseTimeMs() const
 
 
-void FreeBoyInstrument::playNote(NotePlayHandle* nph, SampleFrame* workingBuffer)
+void FreeBoyInstrument::playNoteImpl(NotePlayHandle* nph, CoreAudioDataMut out)
```
Same comment as previous: Why not use `playNote` here?

JohannesLorenz commented on include/AudioPluginInterface.h:
```diff
@@ -0,0 +1,418 @@
+	//virtual void processImpl(NotePlayHandle* nph, ConstBufferT in, BufferT out) {}
+
+	//! The main audio processing method for MIDI-based Instruments
+	virtual void processImpl(ConstBufferT in, BufferT out) = 0;
```
If I were to use this, I would wonder what `in` means for an instrument. Probably, you cannot remove `in` due to design reasons? If so, should a comment be added to clarify that implementers are not allowed to interpret `in` (or are allowed, but only for trivial tests)?

messmerd replied:
The `in` is for instruments with sidechain inputs. If the instrument doesn't have sidechain inputs, `in` will be an empty buffer.

JohannesLorenz replied:
From all I recall, LMMS does not at all support plugins with sidechain inputs. Are you talking about (yet) unreachable code?

messmerd replied:
Sidechain inputs fall into the category of multiple plugin channels, and multiple plugin channels are supported by the pin connector and currently implemented by both VST instruments and VST effects. However since without audio routing there is no way to route a sidechain signal to an instrument's sidechain input, an instrument's sidechain input will always be zeroed. So the functionality is there but not really usable yet.

JohannesLorenz commented on include/lmms_basics.h:
```diff
@@ -74,33 +75,61 @@ constexpr const char* UI_CTRL_KEY =
 /**
  * Simple minimally functional stand-in for C++20's std::span
  *
  * TODO C++20: Use std::span instead
```
Can you add a comment why this cannot be used already now, that we have C++20?

messmerd replied:
We'll have `std::span` once we upgrade to at least GCC 10. I'll add a comment to make that clear

messmerd replied:
Done in https://github.com/LMMS/lmms/pull/7459/commits/3661bbdcc709f3612dd9615d98120326876e3f24

JohannesLorenz commented on include/AudioPluginInterface.h:
```diff
@@ -0,0 +1,418 @@
+		: Instrument{desc, parent, key, flags}
+		, m_pinConnector{config.inputs, config.outputs, this}
+	{
+		connect(Engine::audioEngine(), &AudioEngine::sampleRateChanged, [this]() {
```
There are a lot of `connect` calls, but by default, they are not realtime safe (but even if RT safety does not matter, like here, they are slow). Can you make some of them "direct connections"?

Btw, there have been multiple PRs that transformed some "auto" connections to "direct".

messmerd replied:
I'm considering maybe making `PluginPinConnector` the parent class of the audio plugin buffer class which would remove the need for a bunch of `connect` calls. Currently the pin connector and the plugin's buffer class do not interact with each other directly but instead rely on Qt signals and `AudioPluginInterface` to set up those connections, which may be complicating things.

For example, the plugin buffer initialization for VST plugins is awkward. Currently in `RemotePlugin`, it has to call the pin connector's method to update the channel counts then follow that with an `updateBuffers()` call. Normally calling the pin connector's method to update the channel counts would emit a signal that tells the buffer it needs to update, but this doesn't work while a VST plugin is being initialized (since `m_plugin` is null until initialization is complete, but it needs to be non-null for `AudioPluginBuffer` to access the plugin's buffer to tell it to update itself).

Making the pin connector the base class of the plugin buffer, or coupling the two classes directly together by some other means could solve this issue and remove the need for the `connect` calls.

messmerd replied:
I modified the design as described in my comment above, which removed the connect calls. See: https://github.com/messmerd/lmms/pull/3

JohannesLorenz commented on include/AudioPluginInterface.h:
```diff
@@ -0,0 +1,418 @@
+		 */
+	}
+
+	auto pinConnector() -> PluginPinConnector* { return &m_pinConnector; }
```
Removing this function still allows me to compile (I used Release mode, but I guess this does not matter). I would say let's remove it, to follow the "only expose what's necessary" principle?

messmerd replied:
I think this is needed in `VstEffect::openPlugin()`.

JohannesLorenz replied:
Strange, I now see that it is needed when compiling. I may have compiled without VST support back then.

JohannesLorenz commented on include/AudioPluginInterface.h:
```diff
@@ -0,0 +1,418 @@
+		AudioPluginBufferDefaultImpl<config.layout, SampleT, config.inputs, config.outputs, config.inplace>>
+{
+public:
+	AudioProcessorImpl(const Plugin::Descriptor* desc, Model* parent = nullptr, const Plugin::Descriptor::SubPluginFeatures::Key* key = nullptr)
```
The parent could be made `EffectChain*` here (similar like we use a more concrete class in the Instrument specialization of `AudioProcessorImpl`). However, for weird reasons, our Instruments have code like

```c++
PLUGIN_EXPORT Plugin *lmms_plugin_main(Model *_parent, void *_data)
{
        using KeyType = Plugin::Descriptor::SubPluginFeatures::Key;
        return new Lv2Instrument(static_cast<InstrumentTrack*>(_parent), static_cast<KeyType*>(_data));
        // ...
}
```
but Effects have code like
```c++
PLUGIN_EXPORT Plugin *lmms_plugin_main(Model *_parent, void *_data)
{
        using KeyType = Plugin::Descriptor::SubPluginFeatures::Key;
        return new Lv2Effect(_parent, static_cast<const KeyType*>(_data));
        // ...
}
```
I think it would make sense to use either casts in both plugin mains or in none. I leave it to you to fix or leave it, though it is not really part of this PR.

JohannesLorenz commented on include/Flags.h:
```diff
@@ -52,6 +52,11 @@ class Flags
 	constexpr auto testAny(Flags flags) const -> bool { return (*this & flags) != Flags{}; }
 	constexpr auto testFlag(EnumType flag) const -> bool { return static_cast<bool>(*this & flag); }
 
+	constexpr void setFlag(EnumType flag, bool value = true)
```
This seems to be the only occurrence of  `setFlag` in the PR diff. Is this not used? Or if it is, how could it work before?

messmerd replied:
I was using `setFlag` at one point but I might not anymore. I'll remove it if it isn't being used.

messmerd replied:
Done in https://github.com/LMMS/lmms/pull/7459/commits/3661bbdcc709f3612dd9615d98120326876e3f24

JohannesLorenz commented on include/SampleFrame.h:
```diff
@@ -227,8 +228,56 @@ inline void copyFromSampleFrames(InterleavedSampleType<float>* target, const Sam
+
+
+/**
+ * A non-owning span of CoreAudioData.
```
Adding that this is "interleaved" might make it even more clear.

messmerd replied:
Done in https://github.com/LMMS/lmms/pull/7459/commits/3661bbdcc709f3612dd9615d98120326876e3f24

JohannesLorenz commented on include/SampleFrame.h:
```diff
@@ -227,8 +228,56 @@ inline void copyFromSampleFrames(InterleavedSampleType<float>* target, const Sam
+ * TODO C++23: Use std::mdspan
+ */
+template<typename T>
+struct AudioBus
```
This class is seldomly used, and SampleFrame is included very often. Could we move the definition into PluginPinConnector.h?

messmerd replied:
Done in https://github.com/LMMS/lmms/pull/7459/commits/3661bbdcc709f3612dd9615d98120326876e3f24

JohannesLorenz commented on src/core/audio/AudioPort.cpp:
```diff
@@ -111,21 +111,23 @@ void AudioPort::doProcessing()
 	}
 
 	const fpp_t fpp = Engine::audioEngine()->framesPerPeriod();
+	assert(m_portBuffer.size() == fpp);
```
Why is this done here, but not in so many other places?

JohannesLorenz commented on src/core/audio/AudioPort.cpp:
```diff
@@ -111,21 +111,23 @@ void AudioPort::doProcessing()
 	for( PlayHandle * ph : m_playHandles ) // now we mix all playhandle buffers into the audioport buffer
 	{
-		if( ph->buffer() )
+		if (auto phBuffer = ph->buffer(); phBuffer.data() != nullptr)
```
All the `!= nullptr` in this PR is double negation. What do you consider more readable?
```
(1) If the buffer has anything, ...
(2) If the buffer does not point to nothing, ...
```
I prefer (1) here.

JohannesLorenz commented on src/tracks/InstrumentTrack.cpp:
```diff
@@ -553,9 +557,9 @@ void InstrumentTrack::silenceAllNotes( bool removeIPH )
 f_cnt_t InstrumentTrack::beatLen( NotePlayHandle * _n ) const
 {
-	if( m_instrument != nullptr )
+	if (m_instrument)
```
Here, you handled the double negation the opposite way.

JohannesLorenz commented on include/PluginPinConnector.h:
```diff
@@ -142,22 +198,539 @@ class LMMS_EXPORT PluginPinConnector
+		switch (routedChannels)
+		{
+			case 0b00:
+				// Both track channels are bypassed, so nothing needs to be written to output
```
I think it's not "need", but rather: Nothing is allowed to overwrite the output (in order to preserve the input which shall not be modified)?

messmerd replied:
Done in https://github.com/LMMS/lmms/pull/7459/commits/3661bbdcc709f3612dd9615d98120326876e3f24

JohannesLorenz commented on include/PluginPinConnector.h:
```diff
@@ -142,22 +198,539 @@ class LMMS_EXPORT PluginPinConnector
+				// Both track channels are bypassed, so nothing needs to be written to output
+				break;
+			case 0b01:
+				routeNx2(outPtr, outChannel, std::integral_constant<std::uint8_t, 0b01>{});
```
Actually, you could put all 3 cases together? Even all 4, since the function handles the case of both channels being 0 in a constexpr way.

messmerd replied:
The `std::integral_constant` parameter for each of them is different, so they need to be kept separate

JohannesLorenz replied:
So you use this lambda only to have a `constexpr`, in order to do `constexpr` evaluation? Do I understand this right?

messmerd replied:
It's there so I can use `if constexpr` instead of `if` in any inner loops - mostly for performance reasons. The tradeoff is a slightly larger binary size.

Here's an example and explanation of the concept: https://gist.github.com/messmerd/aef8660b23014ea27e30736c2fb1cf53

Maybe I should use a constexpr array like `HotSpotArray` in the example rather than a switch statement since it might optimize better.

JohannesLorenz replied:
Makes totally sense. I just wasn't sure if I understood your reason. Feel free to keep the code as is or try out the `HotSpotArray` approach. In the end, probably, only a benchmark with a very large for loop will finally tell us what's faster in practice...

JohannesLorenz commented on include/PluginPinConnector.h:
```diff
@@ -142,22 +198,539 @@ class LMMS_EXPORT PluginPinConnector
+
+		if constexpr (rc == 0b11)
+		{
+			// TODO: std::memcpy?
```
Why considering an old C-style function?

JohannesLorenz replied:
But aside from C vs C++, is the comment a reminder that you could add an optimization of using a copy function for the typical case of having exactly 1 input channel for `outPtr`?

messmerd replied:
The comment is here because I wasn't sure how well the compiler would optimize `std::fill_n` vs `std::memcpy`.

This case where both channels need to be zeroed (that is, when `rc == 0b11`) is handled separately not because it needs to be, but because it's an opportunity for optimization. Zeroing every sample in an array with `memcpy` can be done more efficiently than zeroing every other sample such as when only the left channel or only the right channel needs to be zeroed.

JohannesLorenz commented on include/PluginPinConnector.h:
```diff
@@ -142,22 +198,539 @@ class LMMS_EXPORT PluginPinConnector
+				// Normalize output - both channels
+				for (f_cnt_t frame = 0; frame < inOut.frames; ++frame)
+				{
+					outPtr[frame].leftRef() /= numRoutedL;
```
Right now, that I think of it, this division behavior might be strange. Imagine you have a plugin with 3 channels (2 stereo and one for something additional), and want to put it into the 2 core channels, where you put 0->0, 1->1 and 2->0. If the plugin channels 0 and 1 represent stereo audio, after this transformation, the stereo audio will be panned.

This is an exotic case, ofc. Though, I wonder if users should have an option to turn off this behavior? Or even to give weight to the single X markers :smiley:

messmerd replied:
That's a good observation, and I've thought about it myself. The intention of the division behavior was to average the audio in order to make mono plugins sound exactly the same and have the same volume using the pin connector as it would when having two duplicated mono plugin instances (one for left channel, one for right channel), but it creates a problem for plugins with multiple outputs as you've noticed, since a drum machine with 10 outputs will have 1/10th the volume when all 10 outputs are used than when only one output is used.

I talked to @LostRobotMusic about this recently, and we concluded that the pin connector should be summing the channels together rather than averaging them. That is what REAPER's pin connector does.

However, one potentially undesired outcome of this is that mono-input/mono-output plugins connected to both L/R channels of LMMS on their input and output will effectively double the volume of audio passing through it. This is because both of LMMS's track channels are summed together going to the plugin's single input, then the plugin's single output is routed back to both LMMS track channels, so for an effect plugin that does nothing to the audio and is just a passthrough: `Core_L_out = Core_R_out = Core_L_in + Core_R_in` instead the current `Core_L_out = Core_R_out = (Core_L_in + Core_R_in) / 2`.

I could potentially add an option to choose between "averaging" and "summing" behavior rather than making it always summing. If I did that and made "averaging" the default, it would be easier to migrate Lv2 (and maybe LADSPA) away from the current practice of using 2 mono plugin instances to turn mono audio into stereo audio. Though if the "summing" method is unconditionally chosen, maybe the migration could use an Amplifier effect added to the effect chain to compensate for the audio being doubled in volume.

I plan to buy REAPER and do some experimenting to verify that my pin connector behaves identically to the REAPER pin connector it is modeled after.

messmerd replied:
I just bought REAPER and tested it with a mono-input/mono-output plugin.

Here's what it did for the default connections:
![reaper_mono_plugin_default_pin_connector_connections](https://github.com/user-attachments/assets/818cad86-1eea-43dd-b2ae-e50004b97bb6)

It only connects the 1st track channel to the plugin input but connects the plugin output to both track channels. This is a little strange in my opinion, but it does avoid doubling the volume this way.

Interestingly, when I use the "Reset" button, it removes the connection to Track Channel 2 on the output.

messmerd replied:
Fixed in https://github.com/LMMS/lmms/pull/7459/commits/b34df284defdb125d614498c845bb859165ac6c3

JohannesLorenz commented on include/RemotePlugin.h:
```diff
@@ -143,27 +147,26 @@ class LMMS_EXPORT RemotePlugin : public QObject, public RemotePluginBase
+	auto inputBuffer() const -> Span<float> { return m_inputBuffer; }
+	auto outputBuffer() const -> Span<float> { return m_outputBuffer; }
 
+	PluginPinConnector* m_pinConnector = nullptr;
```
I assume these can be `const`, and should be according to the C++ Core Guidelines? (not the pointed-to, but the pointer itself)

This would also be a nice way to express that it's not owning this variable (if it would be owning, a delete-call would conflict with `const`).

messmerd replied:
I'm not aware of the C++ Core Guidelines recommending const member variables in cases like this, but if you point to where it says that, I can change it.

Whenever I use a raw pointer, you can assume it is non-owning. If for some reason I used an owning raw pointer, I would explicitly note that it is owning, otherwise I made a mistake.

JohannesLorenz replied:
The [guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconst-immutable) say "By default, make objects immutable". They mention exceptions, but class members (like in this line) are not mentioned: "Flag non-const variables that are not modified (except for parameters to avoid many false positives and returned local variables)". So I would assume that the pointer should be const. However, you could argue, that like in the "parameters" case, this confuses the reader, making them think the pointed-to memory is immutable.

> Whenever I use a raw pointer, you can assume it is non-owning. If for some reason I used an owning raw pointer, I would explicitly note that it is owning

:+1:

messmerd replied:
Done in https://github.com/LMMS/lmms/pull/7459/commits/3661bbdcc709f3612dd9615d98120326876e3f24

JohannesLorenz commented on plugins/ZynAddSubFx/ZynAddSubFx.h:
```diff
@@ -51,30 +53,39 @@ class LedCheckBox;
 
 
-class ZynAddSubFxInstrument : public Instrument
+// TODO: Is it always 0 inputs, 2 outputs?
```
Why should it not be like that for Zyn?

messmerd replied:
I wasn't sure whether Zyn might have sidechain inputs or extra outputs that we just aren't using

JohannesLorenz replied:
I just asked fundamental, he has been zyn maintainer of zyn since 2009. He says he remembers none, except for the multiple Part outputs, which do not make much sense inside a DAW (they are rather used in MIDI settings). So I conclude the code is fine already.

messmerd replied:
Removed in https://github.com/LMMS/lmms/pull/7459/commits/b34df284defdb125d614498c845bb859165ac6c3

JohannesLorenz commented on plugins/ZynAddSubFx/ZynAddSubFx.h:
```diff
@@ -109,9 +121,12 @@ private slots:
 	ZynAddSubFxRemotePlugin * m_remotePlugin;
 
+	// LocalZynAddSubFx needs to be supplied with a buffer
+	std::optional<AudioPluginBufferDefaultImpl<AudioDataLayout::Split, float, 0, 2, false>> m_localPluginBuffer;
```
Why did you not put this into the `LocalZynAddSubFx` class?

messmerd replied:
This was fixed in https://github.com/messmerd/lmms/pull/3 by adding `ConfigurableAudioPort` which can choose between a local buffer and `RemotePlugin`'s buffer at runtime.

JohannesLorenz commented on plugins/VstEffect/VstEffect.cpp:
```diff
@@ -79,32 +79,44 @@ VstEffect::VstEffect( Model * _parent,
 
 
-Effect::ProcessStatus VstEffect::processImpl(SampleFrame* buf, const fpp_t frames)
+auto VstEffect::processImpl() -> ProcessStatus
```
About your strategy "But also, VSTs and ZynAddSubFX cannot just use any buffer - they use a shared memory buffer.". Couldn't you just create a parameter for `AudioPluginBufferDefaultImpl` to create its buffers in shared memory, rather than "not shared"? That way, you could hide the buffer interface from the Vst/Zyn classes and could use the normal `processImpl(CoreAudioDataMut inOut)`?

messmerd replied:
Maybe, though that would require rewriting a lot more of the `RemotePlugin` class which owns and manages the shared memory, and I'm not sure what the benefit of that would be. It looks like Zyn can run both locally and with `RemotePlugin` which means it needs the option of using `RemotePlugin`'s shared memory as its buffer or providing its own buffer.

Allowing a plugin implementation to provide its own buffer interface (rather than using `AudioPluginBufferDefaultImpl` like most plugins) makes the design really flexible where it needs to be. For example, it would be beneficial to CLAP where a custom buffer implementation would probably be needed to implement its [in-place processing optimization](https://github.com/free-audio/clap/blob/main/include/clap/ext/audio-ports.h#L61:L64).

JohannesLorenz replied:
OK, maybe we leave it like this for now. It's difficult to foresee (or even overview) the different buffer requirements. I'm fine with resolving this.

JohannesLorenz commented on src/core/PluginPinConnector.cpp:
```diff
@@ -27,23 +27,13 @@
 
 #include <QDomDocument>
```
Since this PR changes a lot of includes (and some in headers which are included often): Running include-what-you-use shows some that are not needed (anymore?). However, this PR is "the past", so I will add a comment in the main PR (along with help how to use iwyu).

So, this comment is just for info and can be closed immediately.

JohannesLorenz commented on plugins/Lv2Instrument/Lv2Instrument.h:
```diff
@@ -48,6 +48,7 @@ class Lv2InsView;
 }
 
+// TODO: Add support for pin connector and a variable number of audio input/output ports
 class Lv2Instrument : public Instrument, public Lv2ControlBase
```
Is it correct that none of the Instruments (not even MIDI based ones) have the interface with the new pin connector implemented? I just look for an example, but it seems there is none yet.

Actually, for mono `Lv2Instrument`s, I may even want a pin connector. So maybe we could add this for MIDI-based instruments?

messmerd replied:
Vestige and ZynAddSubFx are both midi-based and support the pin connector, though only Vestige provides a pin connector button so users can access the pin connector GUI. This is because we haven't decided on where the pin connector button should go for instruments.

None of the other instruments support the pin connector yet, though Carla, Lv2, CLAP, and OpulenZ are also midi-based and could potentially add pin connector support in the future. The reason I decided to refrain from giving them pin connector support right now is because the pin connector is not really helpful for instruments unless that instrument supports multiple inputs/outputs and only VST supports that at the moment. Plus the aforementioned issue of no agreed upon location for the pin connector button for instruments.

JohannesLorenz replied:
> Vestige and ZynAddSubFx are both midi-based and support the pin connector

Ah sorry, I meant an Instrument that uses the default buffer implementation, because I was interested in that combination.

Well, in the Lv2 case, it's strictly needed to keep supporting mono plugins in LMMS/lmms#7387 . However, I can delay this (including the button decision) to that PR. However, like mentioned, as a verification measure, I will already try appending a copy of my PR to the pin connector PR to see how well it works.

IMO this comment can be resolved.

messmerd replied:
There aren't any instruments using the default buffer, though OpulenZ could if we wanted to give it pin connector support.

JohannesLorenz commented on include/AudioPluginInterface.h:
```diff
@@ -0,0 +1,418 @@
+{
+
+//! Provides the correct `processImpl` interface for instruments or effects to implement
+template<class ChildT, typename BufferT, typename ConstBufferT, bool inplace, bool customBuffer>
```
A general comment: The whole template logic here is well done and readable for a programmer who is used to templates. However, it seems to me that traditionally, LMMS was always coded with a "keep the code easy and simple" principle, so it might be that in the future, some users might complain that they cannot read the base classes of their plugins anymore. However, this principle might be as deprecated as many others, e.g. the "all Instruments shall be 250x250". Also, while the templates themselves might be hard to read, their use is straightforward (similar to STL code).

So, for me, headers like theses are perfectly fine. However, feel free to ask other developers for more feedback about adding more complex template logic into our "Qt-playground-codebase".

messmerd replied:
Yeah, there may be some opportunities to simplify it further but I'd have to give it some thought

JohannesLorenz replied:
I would only invest time for that if there are actual complaints from users/reviewers. Feel free to close this comment.

JohannesLorenz reviewed (commented):
Functional code review finished :tada: 

Further reviews (style/testing/2nd functional) shall only be done on the base PR IMO.

Technical scope: The scope is only this PR, not the base PR. The base PR will have a separate review.
